### DISCUSSION TITLE
Combobox SelectionChanged handling using ViewModel

### DISCUSSION THREAD

**Discussion by ashah7** [unoplatform/uno#17735] [Q&A] [answered]
Hi

I'm trying to update the value of my combobox when its selection changes using my ViewModel, unsure how it should be done

The combox is populated when the page starts from my `GetEstablishments()` method

Have the below in my Model folder as `EstablishmentOptions.cs`
```

public partial class EstablishmentsOptions : ObservableObject
{
            public int? establishid { get; set; }
            public string description { get; set; }
}
```

My SettingsViewModel.cs

        [ObservableProperty]
        ObservableCollection<EstablishmentsOptions> establishmentsOptions = new ObservableCollection<EstablishmentsOptions>();

A method in my `ViewModel `to Populate my List from value in my MySQL database

```
public void GetEstablishments()

{
    try
    {
        using (MySqlConnection conn = new MySqlConnection(dbconnairmyprayer))
        {
            using (MySqlCommand cmd = new MySqlCommand("getallactiveestablishments", conn))
            {
                cmd.CommandType = System.Data.CommandType.StoredProcedure;
                conn.Open();
                MySqlDataReader myReader = default(MySqlDataReader);
                myReader = cmd.ExecuteReader();
                while (myReader.Read())
                {
                    establishmentsOptions.Add(new EstablishmentsOptions { establishid = myReader.GetInt16(0), description = myReader.GetString(1) } );        
                }
            }
            conn.Close();
            conn.Dispose();
        }
    }
    catch (Exception ex)
    {
        var Message = $"failed to get establishments " + ex.Message;
        Console.WriteLine(Message);
    }
}
```

my xaml in my `Settings.xaml`

`<ComboBox x:Name="comboEstablishmentSelection" Grid.Row="0" Grid.Column="1" FontSize="14" Margin="0,0,0,0"  ItemsSource="{Binding EstablishmentsOptions}" DisplayMemberPath="description" SelectedValuePath="establishid"  SelectedValue="{Binding RequestUserOptions.UserOptions.Establishid}"> </ComboBox>`

**Comment by morning4coffe-dev** [comment, marked as answer]
Hey @ashah7,

I think you are mostly on the correct route, I just updated your code a bit to use the `TwoWay` binding on `SelectedValue`. 

```xml
    <ComboBox x:Name="comboEstablishmentSelection"
              Grid.Row="1"
              HorizontalAlignment="Center"
              VerticalAlignment="Center"
              FontSize="14"
              Margin="0,0,0,0"
              ItemsSource="{Binding EstablishmentsOptions}"
              DisplayMemberPath="description"
              SelectedValuePath="establishid"
              SelectedValue="{Binding RequestUserOptions.UserOptions.Establishid, Mode=TwoWay}" />
```

```csharp
    private int? establishid;
    public int? Establishid
    {
        get => establishid;
        set
        {
            // Work with the selected item as needed
            SetProperty(ref establishid, value);
        }
    }
```
Now the value in the Establishid should update when the selected item in the ComboBox is changed. Let me know if those changes work for you.

**Reply by ashah7** [reply]
thank you, I'll have a go

where is `selectedEstablishmentId `defined?

**Reply by morning4coffe-dev** [reply]
Oh, my bad, it was supposed to be the establishid field defined above the property, I named it differently in my sample to not get myself confused, thanks for pointing this out.

**Reply by ashah7** [reply]
no worries
The two-way binding has taken me in the right direction

I had to put the above code in the `UserOptions.cs ` class for it to fire.  Also managed to find a shorter way for variables decorated in the `[ObservableProperty] `

```
   [ObservableProperty]
   [JsonProperty("establishid")]
   public int? establishid;

 partial void OnEstablishidChanged(int? oldValue, int? newValue)
   {
       if (oldValue == null) { return; }
       Console.WriteLine(oldValue);
       Console.WriteLine("Write to db"+newValue);
   }
```
Using the `OnEstablishidChanged` method I don't need to manually do the getters and setters.

Also unsure how I can update my database because I can't seem to bring in my db connection string from my `appsettings.json` file.  I dont want to use DI because it then puts a parameterised constructor on `UserOptions.cs` class which I don't want.   I'll probably open another question for that one.  Thanks again

**Reply by morning4coffe-dev** [reply]
Nice! Both options should work similarly - either using ObservableProperty with OnChanged or the SetProperty, with the SetProperty you can work with the value directly, but as you mentioned you do have to do the getters and setters. For the other question, please open a new discussion with more details, I will take a look 👍
